### PR TITLE
Json references

### PR DESCRIPTION
**Pull request by ghost**
This is a continuation on draft 4 PR.
It also contains the previous PR with reference support.

This PR adds both local and remote references. It should follow the id-attribute as far as I know. Atleast the basic cases.

### CODE REVIEW

File: test/jesse_tests_draft4_SUITE.erl [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        , ref/1
+        , refRemote/1
+        , required/1
+        , type/1
+        , uniqueItems/1
+        ]).
+
+-include_lib("common_test/include/ct.hrl").
+
+-define(json_schema_draft4, <<"http://json-schema.org/draft-04/schema#">>).
```
Comment (gadgetci) [inline test/jesse_tests_draft4_SUITE.erl]: According to **Elvis**:
> Invalid macro name json_schema_draft4 on line 61. Use UPPER_CASE.

File: src/jesse_json_path.erl [hunk 2]
```diff
@@ -26,6 +26,17 @@
 
 -export_type([proplist/0, kvc_key/0, kvc_obj/0]).
 
+%% @doc Parse a JSON Pointer
+%% TODO: Parse according to specification
+-spec parse(JSONPointer :: string() | binary()) -> [binary()].
+parse(JSONPointer) ->
+    lists:map(
+      fun (Segment) when is_list(Segment) -> unicode:characters_to_binary(Segment);
```
Comment (gadgetci) [inline src/jesse_json_path.erl]: According to **Elvis**:
> Line 34 is too long:       fun (Segment) when is_list(Segment) -> unicode:characters_to_binary(Segment);.

File: test/jesse_tests_draft3_SUITE.erl [hunk 3]
```diff
@@ -216,18 +190,25 @@ run_test_set(Schema, TestSet) ->
...
-                     io:format("Result: ~p~n", [Result]),
-                     case get_path(?VALID, Test) of
-                       true  -> {ok, TestData} = Result;
-                       false -> {error, _} = Result
+                     ct:pal("* Test case: ~s~n", [Description]),
+                     Opts = [{schema_loader_fun, fun load_schema/1}],
+                     try jesse:validate_with_schema(Schema, TestData, Opts) of
+                         Result ->
+                             ct:pal("Result: ~p~n", [Result]),
+                             case get_path(?VALID, Test) of
```
Comment (gadgetci) [inline test/jesse_tests_draft3_SUITE.erl]: According to **Elvis**:
> The expression on line 198 and column 30 is nested beyond the maximum level of 3.

File: src/jesse_state.erl [hunk 4]
```diff
@@ -138,14 +149,120 @@ set_allowed_errors(#state{} = State, AllowedErrors) ->
...
+            _:_ -> ?not_found
+          end;
+        false ->
+          ?not_found
+      end
+  end.
+
+%% @doc Resolve a new id
+%% @private
+-spec combine_id(undefined | http_uri:uri(), undefined | binary()) -> http_uri:uri().
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 228 is too long: -spec combine_id(undefined | http_uri:uri(), undefined | binary()) -> http_uri:uri()..

File: src/jesse_state.erl [hunk 5]
```diff
@@ -35,28 +35,34 @@
...
 -record( state
-       , { original_schema    :: jesse:json_term()
+       , { root_schema        :: jesse:json_term()
          , current_schema     :: jesse:json_term()
          , current_path       :: [binary()] %% current path in reversed order
          , allowed_errors     :: non_neg_integer() | 'infinity'
          , error_list         :: list()
          , error_handler      :: fun((#state{}) -> list() | no_return())
          , default_schema_ver :: atom()
+         , schema_loader_fun  :: fun((binary()) -> {ok, jesse:json_term()} | jesse:json_term() | ?not_found)
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 59 is too long:          , schema_loader_fun  :: fun((binary()) -> {ok, jesse:json_term()} | jesse:json_term() | ?not_found).

File: src/jesse_state.erl [hunk 6]
```diff
@@ -35,28 +35,34 @@
...
+        , resolve_reference/2
         ]).
 
 -export_type([ state/0
              ]).
 
+%% Includes
+-include("jesse_schema_validator.hrl").
+
+-define(schema_loader_fun, fun jesse_database:read/1).
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Invalid macro name schema_loader_fun on line 48. Use UPPER_CASE.

File: test/jesse_tests_draft4_SUITE.erl [hunk 7]
```diff
@@ -0,0 +1,261 @@
+%%%=============================================================================
```
Comment (gadgetci) [inline test/jesse_tests_draft4_SUITE.erl]: According to **Elvis**:
> This module has too many functions (31). Consider breaking it into a number of modules.

File: src/jesse_validator_draft4.erl [hunk 8]
```diff
@@ -0,0 +1,859 @@
...
+    {true, _}  -> handle_data_invalid(?not_schema_valid, Value, State);
+    {false, _} -> State
+  end.
+
+validate_schema(Value, Schema, State0) ->
+  try
+    case jesse_lib:is_json_object(Schema) of
+      true ->
+        State1 = jesse_state:set_current_schema(State0, Schema),
+        State2 = jesse_schema_validator:validate_with_state(Schema, Value, State1),
```
Comment (gadgetci) [inline src/jesse_validator_draft4.erl]: According to **Elvis**:
> Line 739 is too long:         State2 = jesse_schema_validator:validate_with_state(Schema, Value, State1),.

File: test/jesse_tests_draft3_SUITE.erl [hunk 9]
```diff
@@ -52,6 +54,7 @@
 -include_lib("common_test/include/ct.hrl").
 
 -define(TESTS_DIR, "JSON-Schema-Test-Suite/tests/draft3").
+-define(json_schema_draft3, <<"http://json-schema.org/draft-03/schema#">>).
```
Comment (gadgetci) [inline test/jesse_tests_draft3_SUITE.erl]: According to **Elvis**:
> Invalid macro name json_schema_draft3 on line 57. Use UPPER_CASE.

File: test/jesse_schema_validator_tests.erl [hunk 10]
```diff
@@ -179,14 +179,14 @@ empty_list_as_valid_value_for_string_test() ->
...
-  UnsupportedSchema = {[{<<"$schema">>, <<"http://json-schema.org/draft-04/schema#">>}]},
+  UnsupportedSchema = {[{<<"$schema">>, <<"http://json-schema.org/draft-05/schema#">>}]},
 
   Json = {[{<<"Doesn't matter">>}]},
   ?assertEqual( {ok, Json}
               , jesse_schema_validator:validate(SupportedSchema, Json, [])
               ),
   ?assertThrow([{schema_invalid, UnsupportedSchema,
-                 {schema_unsupported, <<"http://json-schema.org/draft-04/schema#">>}}]
+                 {schema_unsupported, <<"http://json-schema.org/draft-05/schema#">>}}]
```
Comment (gadgetci) [inline test/jesse_schema_validator_tests.erl]: According to **Elvis**:
> Line 189 is too long:                  {schema_unsupported, <<"http://json-schema.org/draft-05/schema#">>}}].

File: test/jesse_schema_validator_tests.erl [hunk 11]
```diff
@@ -179,14 +179,14 @@ empty_list_as_valid_value_for_string_test() ->
 
 schema_unsupported_test() ->
   SupportedSchema = {[{<<"$schema">>, <<"http://json-schema.org/draft-03/schema#">>}]},
-  UnsupportedSchema = {[{<<"$schema">>, <<"http://json-schema.org/draft-04/schema#">>}]},
+  UnsupportedSchema = {[{<<"$schema">>, <<"http://json-schema.org/draft-05/schema#">>}]},
```
Comment (gadgetci) [inline test/jesse_schema_validator_tests.erl]: According to **Elvis**:
> Line 182 is too long:   UnsupportedSchema = {[{<<"$schema">>, <<"http://json-schema.org/draft-05/schema#">>}]},.

File: test/jesse_schema_validator_tests.erl [hunk 12]
```diff
@@ -179,14 +179,14 @@ empty_list_as_valid_value_for_string_test() ->
 
 schema_unsupported_test() ->
   SupportedSchema = {[{<<"$schema">>, <<"http://json-schema.org/draft-03/schema#">>}]},
```
Comment (gadgetci) [inline test/jesse_schema_validator_tests.erl]: According to **Elvis**:
> Line 181 is too long:   SupportedSchema = {[{<<"$schema">>, <<"http://json-schema.org/draft-03/schema#">>}]},.

File: test/jesse_tests_draft4_SUITE.erl [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+  lists:foreach( fun(Test) ->
+                     Description = get_path(?DESCRIPTION, Test),
+                     TestData    = get_path(?DATA, Test),
+                     ct:pal("* Test case: ~s~n", [Description]),
+                     Opts = [{default_schema_ver, ?json_schema_draft4},
+                             {schema_loader_fun, fun load_schema/1}],
+                     try jesse:validate_with_schema(Schema, TestData, Opts) of
+                         Result ->
+                             ct:pal("Result: ~p~n", [Result]),
+                             case get_path(?VALID, Test) of
```
Comment (gadgetci) [inline test/jesse_tests_draft4_SUITE.erl]: According to **Elvis**:
> The expression on line 225 and column 30 is nested beyond the maximum level of 3.

File: src/jesse_schema_validator.hrl [hunk 14]
```diff
@@ -80,10 +88,28 @@
...
+-define(wrong_type_specification,    'wrong_type_specification').
+-define(all_schemas_not_valid,       'all_schemas_not_valid').
+-define(wrong_all_of_schema_array,   'wrong_all_of_schema_array').
+-define(any_schemas_not_valid,       'any_schemas_not_valid').
+-define(wrong_any_of_schema_array,   'wrong_any_of_schema_array').
+-define(not_one_schema_valid,        'not_one_schema_valid').
+-define(wrong_one_of_schema_array,   'wrong_one_of_schema_array').
+-define(not_schema_valid,            'not_schema_valid').
+-define(wrong_not_schema,            'wrong_not_schema').
+-define(invalid_schema,              'wrong_schema').
```
Comment (KoFish) [inline src/jesse_schema_validator.hrl]: `invalid_schema` not matching `wrong_schema`

File: src/jesse_validator_draft4.erl [hunk 15]
```diff
@@ -0,0 +1,859 @@
...
+        {true, State2};
+      false ->
+        handle_schema_invalid(?invalid_schema, State0)
+    end
+  catch
+    throw:Errors -> {false, Errors}
+  end.
+
+%% @doc $ref
```
Comment (KoFish) [inline src/jesse_validator_draft4.erl]: The documentation, and use of `@doc` tags, is very poor and should be improved.

File: src/jesse_json_path.erl [hunk 16]
```diff
@@ -26,6 +26,17 @@
 
 -export_type([proplist/0, kvc_key/0, kvc_obj/0]).
 
+%% @doc Parse a JSON Pointer
+%% TODO: Parse according to specification
```
Comment (KoFish) [inline src/jesse_json_path.erl]: Maybe mention something about what from the specification that this function should upon completing TODO? Such as tilde-escaping?

File: src/jesse_validator_draft4.erl [hunk 17]
```diff
@@ -0,0 +1,1248 @@
...
+
+%% @doc Validate a value against a schema in a given state.
+%% Used by all combinators to run validation on a schema.
+%% @private
+validate_schema(Value, Schema, State0) ->
+  try
+    case jesse_lib:is_json_object(Schema) of
+      true ->
+        State1 = jesse_state:set_current_schema(State0, Schema),
+        State2 = jesse_schema_validator:validate_with_state(Schema, Value, State1),
```
Comment (gadgetci) [inline src/jesse_validator_draft4.erl]: According to **Elvis**:
> Line 1127 is too long:         State2 = jesse_schema_validator:validate_with_state(Schema, Value, State1),.

File: src/jesse_json_path.erl [hunk 18]
```diff
@@ -26,6 +26,18 @@
 
 -export_type([proplist/0, kvc_key/0, kvc_obj/0]).
 
+%% @doc Parse a JSON Pointer
+%% TODO: Parse according to specification. This does not accurately parse
+%%       tilde escaped JSON Pointers.
+-spec parse(JSONPointer :: string() | binary()) -> [binary()].
+parse(JSONPointer) ->
+    lists:map(
+      fun (Segment) when is_list(Segment) -> unicode:characters_to_binary(Segment);
```
Comment (gadgetci) [inline src/jesse_json_path.erl]: According to **Elvis**:
> Line 35 is too long:       fun (Segment) when is_list(Segment) -> unicode:characters_to_binary(Segment);.

**Pull request closed without merging**